algebra-cl-gen/src/gpu/sources.rs

```diff
@@ -0,0 +1,264 @@
+    }
+    
+    #[cfg(feature = "tweedle")]
+    if TypeId::of::<G>() == TypeId::of::<tweedle::dee::Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<tweedle::dum::Affine>()
+    {
+        prefix_map.insert(TypeId::of::<tweedle::dee::Affine>(), String::from("Dee_"));
+        prefix_map.insert(TypeId::of::<tweedle::dum::Affine>(), String::from("Dum_"));
+    }
+
+    prefix_map
+}
+
+pub fn kernel_fft<F: PrimeField>(limb64: bool) -> String 
+{
+    vec![
+        if limb64 {
+            field::<F, Limb64>("Fr")
+        } else {
+            field::<F, Limb32>("Fr")
+        },
+        fft("Fr"),
+    ]
+    .join("\n\n")
+}
+
+pub fn kernel_multiexp<G>(limb64: bool) -> String
+where
+    G: AffineCurve
+{   
+    let mut src = String::from("");
+
+    #[cfg(feature = "bn_382")]
+    if TypeId::of::<G>() == TypeId::of::<bn_382::g::Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bn_382::G1Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bn_382::G2Affine>()
+    {
+        src = multiexp_src!(
+            limb64, bn_382, true,
+            "G", "Fr", "Fr",
+            "G1", "Fq", "Fr",
+            "G2", "Fq2", "Fr"
+        );
+    }
+
+    #[cfg(feature = "bls12_381")]
+    if TypeId::of::<G>() == TypeId::of::<bls12_381::G1Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bls12_381::G2Affine>()
+    {
+        src = multiexp_src!(
+            limb64, bls12_381, true,
+            "G1", "Fq", "Fr",
+            "G2", "Fq2", "Fr"
+        );        
+    }
+    
+    #[cfg(feature = "bls12_377")]
+    if TypeId::of::<G>() == TypeId::of::<bls12_377::G1Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bls12_377::G2Affine>()
+    {
+        src = multiexp_src!(
+            limb64, bls12_377, true,
+            "G1", "Fq", "Fr",
+            "G2", "Fq2", "Fr"
+        );        
+    }
+    
+    #[cfg(feature = "bn254")]
+    if TypeId::of::<G>() == TypeId::of::<bn254::G1Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bn254::G2Affine>()
+    {
+        src = multiexp_src!(
+            limb64, bn254, true,
+            "G1", "Fq", "Fr",
+            "G2", "Fq2", "Fr"
+        );        
+    }
+
+    #[cfg(feature = "tweedle")]
+    if TypeId::of::<G>() == TypeId::of::<tweedle::dee::Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<tweedle::dum::Affine>()
+    {
+        src = multiexp_src!(
+            limb64, tweedle, false,
+            "Dee", "Fq", "Fr",
+            "Dum", "Fr", "Fq"
+        );        
+    }
+
+    src
+
+}
+
+pub fn kernel_polycommit<G>(limb64: bool) -> String
+where
+    G: AffineCurve
+{
+    let mut src = String::from("");
+
+    #[cfg(feature = "bn_382")]
+    if TypeId::of::<G>() == TypeId::of::<bn_382::g::Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bn_382::G1Affine>() ||
+       TypeId::of::<G>() == TypeId::of::<bn_382::G2Affine>()
+    {
+        src = polycommit_round_reduce_src!(
+            limb64, bn_382, true,
+            "G", "Fr", "Fr",
```

Comment:
what does the "G", "Fr", "Fr" line mean? Is the bn_382 treated differently than other bn's?